File: src/game/entity/bomb/AbstractExplosion.java
package game.entity.bomb;

import game.Bomberman;
import game.entity.blocks.DestroyableBlock;
import game.entity.models.*;
import game.ui.panels.game.PitchPanel;
import game.values.DrawPriority;

import java.awt.image.BufferedImage;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static game.utils.Utility.loadImage;

/**
 * An abstract class for in-game explosions;
 */
public abstract class AbstractExplosion extends MovingEntity {
    public static int MAX_EXPLOSION_LENGTH = 5;
    public static final int SIZE = PitchPanel.COMMON_DIVISOR * 2;
    public static final int SPAWN_OFFSET = (PitchPanel.GRID_SIZE-SIZE) / 2;
    // The distance from the bomb where the explosion was created.
    protected final int distanceFromExplosive;

    // The maximum distance from the bomb that the explosion can travel.
    protected final int maxDistance;
    protected static final int BOMB_STATES = 3;
    protected boolean canExpand;

    // The direction of the explosion.
    protected final Direction direction;
    protected boolean appearing = true;
    protected int explosionState = 1;
    protected long lastRefresh = 0;
    private final Explosive explosive;
    private ExplosiveCaller caller;

    public AbstractExplosion(Coordinates coordinates, Direction direction, Explosive explosive) {
        this(coordinates, direction, 0, explosive);
    }

    public AbstractExplosion(Coordinates coordinates, Direction direction, Integer distanceFromBomb, Explosive explosive){
        this(coordinates,direction,distanceFromBomb,explosive,true);
    }

    /**
     * Constructs a new explosion.
     *
     * @param coordinates           The starting coordinates of the explosion.
     * @param direction             The direction of the explosion.
     * @param distanceFromExplosive The distance from the bomb where the explosion was created.
     */
    public AbstractExplosion(Coordinates coordinates, Direction direction, Integer distanceFromExplosive, Explosive explosive, boolean canExpand) {
        super(coordinates);

        this.direction = direction;
        this.distanceFromExplosive = distanceFromExplosive;
        this.explosive = explosive;
        this.maxDistance = explosive.getMaxExplosionDistance();
        this.canExpand = canExpand;

        //on first (center) explosion
        if (distanceFromExplosive == 0) {
            List<Coordinates> desiredCoords = getAllCoordinates();
            Bomberman.getMatch().getEntities()
                    .parallelStream()
                    .filter(e -> desiredCoords.stream().anyMatch(coord -> Coordinates.doesCollideWith(coord, e)))
                    .forEach(this::interact);
        }

        if (getCanExpand())
            moveOrInteract(direction, getSize(), true);
    }

    @Override
    public DrawPriority getDrawPriority() {
        return DrawPriority.DRAW_PRIORITY_4;
    }

    /**
     * Interacts with another entity in the game.
     *
     * @param e The entity to interact with.
     */
    @Override
    protected void doInteract(Entity e) {
        if (e instanceof BomberEntity || e instanceof Enemy || e instanceof DestroyableBlock) {
            attack(e);
        } else if (e instanceof Bomb) {
            ((Bomb) e).explode();
        }
    }

    /**
     * Returns the size of the explosion.
     *
     * @return The size of the explosion.
     */
    @Override
    public int getSize() {
        return SIZE;
    }

    /**
     * Sets the coordinates of the explosion and creates new explosions based on its distance from the bomb.
     *
     * @param coordinates The new coordinates of the explosion.
     */
    @Override
    public void move(Coordinates coordinates) {
        Coordinates nextTopLeftCoords = nextCoords(direction, getSize());

        try {
            Constructor<? extends AbstractExplosion> constructor = getExplosionClass().getConstructor(Coordinates.class, Direction.class, int.class, Explosive.class);

            constructor.newInstance(
                    nextTopLeftCoords,
                    direction,
                    distanceFromExplosive + 1,
                    getExplosive()
            ).spawn(true,false);
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            e.printStackTrace();
        }
    }

    public int getState() {
        if (explosionState == 0 && !appearing) {
            despawn();
            appearing = true;
            return 0;
        }

        if (explosionState == BOMB_STATES) appearing = false;

        int appearingConstant = !appearing ? -1 : 1;

        int prevState = explosionState;
        long currentTime = System.currentTimeMillis();

        if (currentTime - lastRefresh >= getImageRefreshRate()) {
            explosionState += appearingConstant;
            lastRefresh = currentTime;
        }

        return prevState;
    }

    @Override
    public int getImageRefreshRate() {
        return 100;
    }

    @Override
    public boolean isObstacle(Entity e) {
        return (e == null) || getExplosive().isObstacleOfExplosion(e);
    }

    @Override
    public Set<Class<? extends Entity>> getObstacles() {
        return new HashSet<>(getExplosive().getExplosionObstacles());
    }

    @Override
    public Set<Class<? extends Entity>> getInteractionsEntities(){
        return new HashSet<>(getExplosive().getExplosionInteractionEntities());
    }

    @Override
    public BufferedImage getImage() {
        if (distanceFromExplosive == 0) {
            return loadImage(String.format("%s_central" + getState() + ".png", getBasePath()));
        }

        String isLast = canExpand ? "" : "_last";
        String imageFileName = "_" + direction.toString().toLowerCase();

        // Load and set the image of the flame.
        String imagePath = String.format("%s%s%s%s.png", getBasePath(), imageFileName, isLast, getState());
        return loadAndSetImage(imagePath);
    }

    public boolean getCanExpand() {
        if (distanceFromExplosive >= maxDistance) canExpand = false;
        return canExpand;
    }
    public ExplosiveCaller getCaller(){
        if(explosive instanceof ExplosiveCaller){
            return (ExplosiveCaller) explosive;
        }if (explosive instanceof CalledExplosive){
            return ((CalledExplosive) explosive).getCaller();
        }
        return null;
    }

    public void onObstacle(Coordinates coordinates){
        try {
            Constructor<? extends AbstractExplosion> constructor = getExplosionClass().getConstructor(Coordinates.class, Direction.class, int.class, Explosive.class, boolean.class);

            constructor.newInstance(
                    coordinates,
                    direction,
                    distanceFromExplosive + 1,
                    explosive,
                    false
            ).spawn(true,false);
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            e.printStackTrace();
        }
    }

    public Explosive getExplosive() {
        return explosive;
    }

    protected abstract Class<? extends AbstractExplosion> getExplosionClass();

    @Override
    protected Set<Class<? extends Entity>> getBasePassiveInteractionEntities() {
        return new HashSet<>();
    }
}

File: src/game/entity/bomb/AbstractExplosion.kt
package game.entity.bomb

import game.Bomberman
import game.entity.models.*
import game.ui.panels.game.PitchPanel
import game.utils.Utility
import game.values.DrawPriority
import java.awt.image.BufferedImage
import java.lang.Exception
import java.util.*

/**
 * An abstract class for in-game explosions;
 */
abstract class AbstractExplosion(private val owner: Entity,
                                 coordinates: Coordinates?,
                                 protected val direction: Direction,
                                 private val distanceFromExplosive: Int,
                                 val explosive: Explosive,
                                 private var canExpand: Boolean
) : MovingEntity(coordinates) {
    private var appearing = true
    private var explosionState = 1
    private var lastRefresh: Long = 0

    constructor(owner: Entity,
                coordinates: Coordinates?,
                direction: Direction,
                explosive: Explosive) : this(owner, coordinates, direction, 0, explosive)

    constructor(owner: Entity,
                coordinates: Coordinates?,
                direction: Direction,
                distanceFromBomb: Int,
                explosive: Explosive) : this(owner, coordinates, direction, distanceFromBomb, explosive, true)

    init {
        // On first (center) explosion
        if (distanceFromExplosive == 0) {
            val desiredCoords = allCoordinates
            Bomberman.getMatch().getEntities()
                    .parallelStream()
                    .filter { e: Entity -> desiredCoords.any { coord: Coordinates? -> Coordinates.doesCollideWith(coord, e) } }
                    .forEach { e: Entity -> interact(e) }
        }

        if (canExpand()) {
            expandBomb(direction, size)
        }
    }

    /**
     * Returns the most specific class of the explosion instance
     */
    protected abstract val explosionClass: Class<out AbstractExplosion>

    /**
     * Returns the size of the explosion.
     *
     * @return The size of the explosion.
     */
    override fun getSize(): Int = SIZE

    override fun getDrawPriority(): DrawPriority = DrawPriority.DRAW_PRIORITY_4

    override fun getBasePassiveInteractionEntities(): Set<Class<out Entity>> = HashSet()

    override fun getImageRefreshRate(): Int = 100

    override fun isObstacle(e: Entity?): Boolean = e == null || explosive.isObstacleOfExplosion(e)

    override fun getObstacles(): Set<Class<out Entity>> = explosive.explosionObstacles

    override fun getInteractionsEntities(): Set<Class<out Entity>> = explosive.explosionInteractionEntities

    private val _state: Int
        get() {
            if (explosionState == 0 && !appearing) {
                despawnAndNotify()
                appearing = true
                return 0
            }

            if (explosionState == BOMB_STATES)
                appearing = false

            val appearingConstant = if (!appearing) -1 else 1
            val prevState = explosionState

            if (Utility.timePassed(lastRefresh) >= imageRefreshRate) {
                explosionState += appearingConstant
                lastRefresh = System.currentTimeMillis()
            }

            return prevState
        }

    override fun getImage(): BufferedImage {
        val imageName = when (distanceFromExplosive) {
            0 -> {
                val centralImage = "_central$_state.png"
                "${entitiesAssetsPath}$centralImage"
            }

            else -> {
                val isLast = if (canExpand) "" else "_last"
                val imageFileName = "_${direction.toString().lowercase()}"
                val expandedImage = "${entitiesAssetsPath}$imageFileName$isLast${_state}.png"
                expandedImage
            }
        }

        return loadAndSetImage(imageName)
    }

    private fun canExpand(): Boolean {
        if (distanceFromExplosive >= explosive.maxExplosionDistance) canExpand = false
        return canExpand
    }

    /**
     * Returns whether it should hide the central entity or not.
     */
    protected open fun shouldHideCenter(): Boolean = false

    /**
     * Interacts with another entity in the game.
     *
     * @param e The entity to interact with.
     */
    override fun doInteract(e: Entity?) {
        e?.onExplosion(this)
    }

    /**
     * Sets the coordinates of the explosion and creates new explosions based on its distance from the bomb.
     *
     * @param coordinates The new coordinates of the explosion.
     */
    override fun move(coordinates: Coordinates) {
        val nextTopLeftCoords = Coordinates.nextCoords(coords, direction, size)
        try {
            val constructor = explosionClass.getConstructor(
                    Entity::class.java,
                    Coordinates::class.java,
                    Direction::class.java,
                    Int::class.javaPrimitiveType,
                    Explosive::class.java
            )
            constructor.newInstance(
                    owner,
                    nextTopLeftCoords,
                    direction,
                    distanceFromExplosive + 1,
                    explosive
            )?.explode()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun onObstacle(coordinates: Coordinates?) {
        try {
            val constructor = explosionClass.getConstructor(
                    Entity::class.java,
                    Coordinates::class.java,
                    Direction::class.java,
                    Int::class.javaPrimitiveType,
                    Explosive::class.java,
                    Boolean::class.javaPrimitiveType
            )
            constructor.newInstance(
                    owner,
                    coordinates,
                    direction,
                    distanceFromExplosive + 1,
                    explosive,
                    false
            )!!.explode()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun expandBomb(d: Direction, stepSize: Int) {
        moveOrInteract(d, stepSize, true)
    }

    fun explode() {
        spawn(true, false)
    }

    companion object {
        val SIZE = PitchPanel.COMMON_DIVISOR * 2
        val SPAWN_OFFSET = (PitchPanel.GRID_SIZE - SIZE) / 2
        var MAX_EXPLOSION_LENGTH = 5
        protected const val BOMB_STATES = 3
    }
}